### PR TITLE
added api routes

### PR DESCRIPTION
**Pull request by dmbf29**
So actually I'm unsure the main page of the app- the page where the user sees all of the matches and their predictions (if any).
 
My first instinct was to just have it at `/matches` because if a user is logged in, easy to get the predictions for the matches, right? 
But then, if you want to see other people's predictions (maybe after they're locked) then the routes definitely **could not** be `/matches` for them. Unless you passed user_id in the params or something.
So i'm thinking that it could be at `users/:id` since that's basically all you need to see about a user anyway. or you could have `predictions/users/:id`, but then it seems odd to have two separate routes basically handling the same thing, right?

But also, we need to scope this in terms of the competition right? we can't just do `/matches` since we plan to add more competitions later on.

To summarize:
- I nested routes inside of `/competitions/:id` (except for ones that don't need nesting) 
- i left `competitions/:id/matches` in there but I feel like we should use the same route for both the logged in user + other users.
- something doesn't feel right with `/competitions/:id/users/:id`, i feel like there's a better way to do it

#11

### CODE REVIEW

**Comment by dmbf29** [comment]
Whoops I pushed an unsaved version originally 😅

**Comment by trouni** [comment]
Ok so I've been scratching my head over this for a little while now, and here are my thoughts:
- Instead of nesting the routes under `competitions`, I think we should nest them under `users` => one of the reasons is that it would make sense to want to watch a user's profile and see all of the competitions or leagues they took part in.
- I think `competitions` could be made as an optional query param for filtering => technically, if multiple competitions were to happen at the same time, you could still list all of the matches (the competition they belong to would show up as well in the list) and make predictions for all of them (if you're a heavy user and participate in many competitions). You could of course want to filter by competition, but that would become technically optional.
- Thinking of the reverse option, `competitions/:id/matches?user_id=:id` would not really make sense because I can't think of a situation where we would want to see all the matches/predictions for all app users for a single competition.

So in the end:
- `users/:id/matches` to see all matches and predictions for any user
- `users/:id/matches?competition_id=:id` to filter those matches for a specific competition. This approach would also allow more specific filtering like `users/:id/matches?round_id=:id` or even `users/:id/matches?group_id=:id` (not very useful, but I think it helps with wrapping our heads around the logic)

What do you think?

**Comment by dmbf29** [comment]
I like that. However, a user will have be logged in to see matches. 
Perhaps, we should leave `competitions/:id/:matches` for a "visitor" then?

**Comment by trouni** [comment]
Good point... If we follow the same logic, then we could have a `/matches` route and filter by competition with `/matches?competition_id=:id`.

That raises the question of should we have everything as parameters? `/matches?competition_id=:id&user_id=:id`?

**Comment by dmbf29** [comment]
lol most complicated routes ever.

Let's go with the parameter routes. Updated now. (Also updated [routes sheet])(https://docs.google.com/spreadsheets/d/12BLfUduioi7LzVv7m0ORii-uQbRd9-xjP5TTlGbSok0/edit#gid=0)

**Review by trouni — APPROVED** [review]
Yeah the more I think about it, the more I think the parameter approach is the way to go. We're retrieving `matches`, the rest are optional filters.

File: config/routes.rb [hunk 1]
```diff
@@ -1,3 +1,17 @@
...
+      resources :leagues, only: [:index, :create]
+      resources :users, only: [:show]
+      resources :memberships, only: [:create]
+    end
+    resources :matches, only: [] do
+      resources :predictions, only: [:create]
+    end
+    resources :predictions, only: [:update]
+    resources :leagues, only: [:destroy]
+    resources :memberships, only: [:destroy]
```
[the remark this answers is not in the document]
Comment (dmbf29) [reply config/routes.rb]: well that's neat 👍. It's not very 'ruby like' i feel like. not totally obvious on the eyes. but made the fix anyway. cleaner this way.
